addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-upload-file-watcher.go

```diff
@@ -2,132 +2,295 @@
+	_, err := svc.DeleteObject(deleteObjectInput)
+
+	if err == nil {
+		if debug {
+			log.Println("Successfully deleted", filename, "from", bucket+"/"+fileKey)
+		}
+	} else {
+		log.Println("Failed to delete object: ", err)
+	}
+
+	return err
+}
+
+func uploadToS3(sess *session.Session, syncDir string, filename string, bucket string, prefix string, kmsKeyId string, debug bool) error {
 	file, err := os.Open(filename)
 	if err != nil {
 		log.Println("Unable to open file", err)
 		return err
 	}
 	defer file.Close()
-	key := strings.TrimPrefix(filename, syncdir)
 	uploader := s3manager.NewUploader(sess)
-	_, err = uploader.Upload(&s3manager.UploadInput{
-		Bucket:               aws.String(bucket),
-		Key:                  aws.String(prefix + "/" + filepath.ToSlash(key)),
-		Body:                 file,
-		ServerSideEncryption: aws.String("aws:kms"),
-		SSEKMSKeyId:          aws.String(kmsKeyId),
-	})
+
+	fileKeyInS3 := ToS3KeyForFile(filename, prefix, syncDir)
+
+	// Do NOT upload if there is no change in file size (bytes)
+	// Without this there will be infinite loop between the downloader thread and the upload watcher thread as follows
+	// Say, the file watcher is watching directory "A", the downloader thread syncs all files from S3 to "A"
+	// This will trigger the file watcher events, the file watcher will upload them (if we don't check size)
+	// The upload in S3 will cause the file's ETag to change even though there is no change in file's content
+	// Due to this, the downloader thread will detect this as file update in S3 and download the file again
+	// This will cause file change event in file watcher and so on...
+	if areSizesDifferent(sess, bucket, fileKeyInS3, file) {
+		var uploadInput *s3manager.UploadInput
+		if strings.TrimSpace(kmsKeyId) == "" {
+			uploadInput = &s3manager.UploadInput{
+				Bucket:               aws.String(bucket),
+				Key:                  aws.String(fileKeyInS3),
+				Body:                 file,
+				ServerSideEncryption: aws.String("aws:kms"),
+			}
+		} else {
+			uploadInput = &s3manager.UploadInput{
+				Bucket:               aws.String(bucket),
+				Key:                  aws.String(fileKeyInS3),
+				Body:                 file,
+				ServerSideEncryption: aws.String("aws:kms"),
+				SSEKMSKeyId:          aws.String(kmsKeyId),
+			}
+		}
+
+		// upload file to S3
+		_, err = uploader.Upload(uploadInput)
+
+		if err == nil {
+			if debug {
+				log.Println("Successfully uploaded", filename, "to", bucket+"/"+fileKeyInS3)
+			}
+		} else {
+			log.Println("Unable to upload", filename, bucket, err)
+		}
+
+	} else {
+		if debug {
+			log.Println(filename, " size has not changed since last upload, skipping upload this time")
+		}
+	}
+
+	return nil
+}
+
+// Checks if the file's sizes are different on disk and in S3
+func areSizesDifferent(sess *session.Session, bucket string, fileKeyInS3 string, file *os.File) bool {
```

Comment:
Was looking at `aws s3 sync` documentation and they also include the criteria of the last modified time of the local file being newer than last modified time of the s3 object. Should that be added as well? If the size is same but the content is different, we shouldn't upload the file if it is older than one in S3.

Reply:
I don't think it is necessary in our case because of the following reasons,
1. A file is uploaded to S3 only if there is CREATE or WRITE event at file system level. So the only reason why we attempt to upload is because either the file is created locally or modified locally and we received event from file system. So the case of local file being older than S3 and us trying to upload never happens.
2. If the file is modified directly on S3 (hence leaving us with older version on file system), the file on file system will be updated during the periodic download sync

In case of `aws s3 sync` it may be necessary to check file timestamps because the user specifies a source and a target and sync is explicitly triggered by the user and not happening in reaction to file system events. So a user may call `aws s3 sync` and specify a local directory as source and S3 location as target and the source may contain files older than the ones in target.